### ISSUE TITLE
Improve backup and restore capabilities

### ISSUE THREAD

**Issue by da2x** [flatpak/flatpak#1356]
TL;DR: It’s hard to backup a flatpak installation. There should be better tools and documentation for backing up and restoring installed repositories and apps.

Backing up user-data is as easy as backing up `$HOME/.var/app/` and excluding `$HOME/.var/app/cache/`. However, backing up just enough data to restore the installed repositories and the reinstall the previously installed apps is more complicated.

It’s undesirable to backup all of `$HOME/.local/share/flatpak/`. The directory can contain gigabytes of data that can be readily downloaded again from the web. Backing up everything can take a lot of time and can cost quite a bit.

## Repository data

Backing up the `$HOME/.local/share/flatpak/repo/config` files and `$HOME/.local/share/flatpak/repo/*.trustedkeys.gpg` files, and copying them back should be enough to restore repositories. However, flatpak won’t work if you just restore these files. `flatpak update` will complain about missing directories ( `repo/objects/` and `repo/refs/remotes/`). If you manually recreate these missing directories, then flatpak will be able to update appdata but won’t be able to use them with any other command (such as install) without getting the error message `error: Remote "example" not found`.

With the suggested patch in pull request #1354, you can export the list of repositories to a set of commands that can be used to recreate the repositories to working condition:

    flatpak remotes --show-details | awk '{print "echo \"echo \\\x22$(base64 \
    --wrap=0 < $HOME/.local/share/flatpak/repo/" $1 ".trustedkeys.gpg)\\\x22 \
    | base64 -d | flatpak remote-add --if-not-exists --gpg-import=- \
    --prio=\\\x22"$4"\\\x22 --title=\\\x22"$2"\\\x22 --user \\\x22"$1"\\\x22 \
    \\\x22"$3"\\\x22\""}' | sh

Backing up and restoring `$HOME/.local/share/flatpak/repo/config` and `$HOME/.local/share/flatpak/repo/*.trustedkeys.gpg` should be enough on its own to restore repositories.

Alternatively, there should be a `flatpak remotes --export` option or something that exports a `local.flatpakrepo` containing all the local repositories.

## Reinstalling apps

With the suggested patch in pull request #1354, you can also export a list of commands that let you reinstall all apps (and --assumeyes to also install any required runtimes):

    flatpak list --app --show-details | awk '{print "flatpak install \
    --assumeyes --user \""$2"\" \""$1}' | cut -d "/" -f1 | awk '{print $0"\""}'

I’m not sure how to improve on this other than providing a `flatpak list --app --export` tool.

**Comment by Kaurin** [comment]
Just to add one thing here: You mentioned that the configs are easily backed up by grabbing "$HOME/.var/app/". I don't think that's the case.

I had installed Atom from flathub, and tried to put my existing backed-up configs into `$HOME/.var/app/io.atom.Atom/config/Atom` . Starting the Flatpak Atom completely ignored by configs because it starts in the sandbox mode. Fair enough.

What I did then was exec into it like so: `flatpak run -d --command=/usr/bin/bash io.atom.Atom/x86_64/stable`, and then copy the config dir into the sandbox encrypted mount, but restarting the Atom resets the config again.

TLDR: Sanboxed apps lack an easy way to backup/restore configs from/to.

I do hope that I'm horribly wrong, and that there is an easy way to do this, but so far I came up with nothing.

I switched back to the RPM Atom.

**Comment by TingPing** [comment]
> Just to add one thing here: You mentioned that the configs are easily backed up by grabbing "$HOME/.var/app/". I don't think that's the case.
> 
> I had installed Atom from flathub, and tried to put my existing backed-up configs into `$HOME/.var/app/io.atom.Atom/config/Atom` . Starting the Flatpak Atom completely ignored by configs because it starts in the sandbox mode. Fair enough.

As in you moved configs from `~/.config/Atom`? That would make sense to fail since relocating configs is usually undefined. But moving configs from the same path to the same path should be fine.

**Comment by Kaurin** [comment]
I have moved the existing configs from `~/.config/Atom`, yes (`.atom` only contains packages, which is unimportant as I have a list of them to restore via apm).

Whay would that make sense to fail? What does "relocating configs is usually undefined" mean?

How are people to restore configs, even from flatpak?

**Comment by TingPing** [comment]
> Whay would that make sense to fail? What does "relocating configs is usually undefined" mean?

Because an application very often writes full paths in its config files, cache files, etc and moving them now makes all of those configurations nonsense.

At the end of the day literally nothing about application configs have anything to do with Flatpak. Applications store whatever they want in those files and if they do or do not like being moved around is up to them.

**Comment by Kaurin** [comment]
I don't think that's the reason why. Something else is a problem here, and here is proof:

1. Install fresh Atom from flatpak.
1. Open a project directory and a file so as to test viewpoort state
1. Turn off "Automatically Update" in preferences so as to test config state change
1. Quit flatpak Atom
1. Backup Flatpak Atom: `cp -r ~/.var/app/io.atom.Atom/config/Atom ~/tmp/` (probably incorrect/irrelevant)
1. Remove config: `rm -rf ~/.var/app/io.atom.Atom/config/Atom`
1. Start Flatpak Atom. View/Configs are untouched. This is to be expected because of sandbox/internal mounts
1. Exit the Flatpak Atom
1. Try deleting the config dir from within the sandbox:
1. `flatpak run -d --command=/usr/bin/bash io.atom.Atom/x86_64/stable`
1. `rm -rf ~/.var/app/io.atom.Atom/config/Atom`
1. `rm -rf /var/config/Atom`
1. `exit` - Exits sanbox shell
1. Start flatpak Atom. View/Configs are still untouched, even though I deleted the only two directories that could have possibly kept those configs.
1. Stop flatpak Atom
1. Go back into the sandbox environment. Both `~/.var/app/io.atom.Atom/config/Atom` and `/var/config/Atom` Exist again.


I'm perplexed at this strange immutability. Where is this behavior documented?

**Comment by Kaurin** [comment]
Update: I tried the same steps above without the `-d` parameter (development runtime). Same results

**Comment by TingPing** [comment]
I don't understand what the problem is you are trying to explain. Step 14 starts Atom, of course they exist by step 16.

**Comment by Kaurin** [comment]
But I have deleted the config dirs within the sandbox in steps 10-12

**Comment by TingPing** [comment]
Most applications make config files on start or exit.

**Comment by Kaurin** [comment]
They don't remake the settings I had specifically configured: Creating a viewport and unticking some options. Those were perserved

**Comment by Kaurin** [comment]
Here, look. Try this yourself and let me know if this makes sense:

1. Install fresh Atom from flatpak.
1. Open a project directory and a file so as to test viewpoort state
1. Turn off "Automatically Update" in preferences so as to test config state change and/or mess with other options
1. Quit flatpak Atom
1. Try deleting the config dir from within the sandbox:
1. `flatpak run -d --command=/usr/bin/bash io.atom.Atom/x86_64/stable`
1. `rm -rf ~/.var/app/io.atom.Atom/config/Atom`
1. `rm -rf /var/config/Atom`
1. `exit` - Exits sanbox shell
1. Start flatpak Atom. View/Configs are still untouched, even though I deleted the only two directories that could have possibly kept those configs.

**Comment by rugk** [comment]
Any news on the original issue?

At least a bulk import & export of all installed apps would really be required. I mean, apt, dnf etc. can all do this, so you should too!

**Comment by rugk** [comment]
And BTW, I think the better (new) example for backing up the list of installed apps may be this?

```
$flatpak list --app --columns=ref| \
awk '{print "flatpak install --assumeyes --user " $1}'
flatpak install --assumeyes --user org.mozilla.FirefoxDevEdition/x86_64/master
[…]
```

**Comment by kirbyfan64** [comment]
Super late to this, but @Kaurin are sure sure Atom closed fully? Some editors like to stay running in the background, and when you "start" it you just bring the previous instance to the foreground.

But I'd argue to backup apps you might as well backup the entire OSTree repository, I think rsync can back up hardlinks properly.

**Comment by rugk** [comment]
> I'd argue to backup apps you might as well backup the entire OSTree repository

Seriously no. This would be equivalent to backing up all the `/usr` and all libs and software data in a usual dnf/rpm installation.

As such, your backup would contain possibly hundreds of GB, which could always be downloaded from the internet if you restore things later.
Even if you backup hardlinks properly, you still have at least one copy of the "useless" data inside of your backup.

E.g. contrast backing up a flatpak installation (explained above) to e.g., backing up dnf:
```sh
dnf list installed > "/path/to/backup/dnf.list" 2>/dev/null
rpm -qa > "/path/to/backup/rpm.list" 2>/dev/null
```

This is all you need for a proper backup of dnf/rpm!

**Comment by TingPing** [comment]
@rugk 	`flatpak list --columns=origin,ref` is basically equal to that.

**Comment by rugk** [comment]
Thanks, but in my test this does not really seem to work:
```sh
$flatpak list --columns=origin,ref | flatpak install
Fehler: At least one REF must be specified

See 'flatpak install --help'
```

**Comment by resolritter** [comment]
Bash scripts (I'm using Bash 1.8 if that matters)

Export: `flatpak list --user --app --columns=origin,ref | tail -n +1 > export`

Install: `while read origin pkg; do flatpak install --user --assumeyes $origin $pkg; done < export`

**Comment by rugk** [comment]
@resolritter Looks good, though maybe also use it without `--user` to backup system-installed flatpaks. And well… if one needs to get the "this was installed as system and this as user" status back properly, one has two commands and two files to use when backing up/restoring.

**Comment by securerootd** [comment]
> Bash scripts (I'm using Bash 5.1.8 if that matters)
> 
> Export: `flatpak list --user --app --columns=origin,ref | tail -n +1 > export`
> 
> Install: `while read origin pkg; do flatpak install --user --assumeyes $origin $pkg; done < export`

`flatpak list --columns=origin,application --app` is probably a better idea as it will filter out the runtimes
